### DISCUSSION TITLE
MIDI Learn

### DISCUSSION THREAD

**Discussion by mikefix1** [RomanKubiak/ctrlr#690] [Ideas]
When adding a new slider or editing an old one, I'd love to have sliders learn what the CC is by turning the knob on the synth.  This would make assigning sliders go faster.

**Comment by damiensellier** [comment]
Hi, do you need at the ctrlr level or panel level?
You would need to add a button "learn" to switch to learn mode. It would trigger a midi learn LUA script "when midi message is received" to catch the cc index and assign it to the clicked/selected slider you want.

**Reply by mikefix1** [reply]
What I'm suggesting is, when you add a slider to a panel, you press a learn button, and then turn the knob or slider on the synth you're making the panel for, and then the slider on the panel learns what the MIDI CC is so you don't need to go and assign it.  This would help speed up the process of making a panel.  Unless I'm missing something that makes it easier (totally possible)?

**Reply by dobo365** [reply]
Well, it is a nice idea but actually, as-is, you just need to indicate it is a CC message and type its number...
Not sure you are gaining a lot of time, isn't it?
Nice to have, definitively.

**Comment by dnaldoog** [comment]
Maybe something like this for an idea?

<pre>
midiReceived = function(--[[ CtrlrMidiMessage --]] midi)
    local s = midi:getSize()
    if s == 3 then
        if _c.setListener:getValue() > -1 then
            local ccNumber = midi:getData():getByte(1)
            local t = {"learn1", "learn2"}
            _c(t[_c.setListener:getValue() + 1]):getMidiMessage():setProperty("midiMessageCtrlrNumber", ccNumber, true)
        end
    end
end
</pre>

The `_c` table is a shortcut to load Ctrlr variables more quickly.

[MIDI Learn_2_0.zip](https://github.com/user-attachments/files/17880008/MIDI.Learn_2_0.zip)

**Reply by damiensellier** [reply]
Hi John,yep exactly! I was thinking about something like this. Right-click or command-click on the mod could also be a trigger for the midi learn process to avoid selecting the mod in the combo and the on/off.
Kudos for the neat script.

**Comment by dnaldoog** [comment]
Hi Damien!

Thank you! 

That's a great idea.

I have made version 3.0 which uses Ctrl/Cmd click - would be interested to know if Cmd click works in macOs.

Regards,

![{7330146F-710A-439F-9A25-B3B62F7CB1C7}](https://github.com/user-attachments/assets/596b1751-492c-40d9-a075-2d34f1cba85b)

[MIDI Learn_3_0.zip](https://github.com/user-attachments/files/17891406/MIDI.Learn_3_0.zip)

**Reply by damiensellier** [reply]
Hi John, I just tried on my macbook and it's working with cmd + click.
Congrats!

**Reply by dnaldoog** [reply]
That’s great to hear. I guess the OP has moved on to greener pastures 😅

**Reply by mikefix1** [reply]
I’m here!   I just haven’t had a chance to try this but it looks great!!!   Will try tomorrow (Tuesday).  Thank you!!

**Reply by dnaldoog** [reply]
Lurking in the shadows- no pressure 
😅

**Reply by mikefix1** [reply]
This worked great in version 5.3.198.  In version 5.6.23 it gave me an error, 
At line [11]: [string "midiReceived"]
Error message: [string "midiReceived"]:11: attempt to call method 'setProperty' (a nil value)
Method disabled. 

Since it works in 5.3.198, all good!  This will help build panels quicker and not have to reference any manuals, etc.  

THANK YOU!!!

**Comment by dnaldoog** [comment]
That’s great to hear. @damiensellier 
can probably tell you why that method doesn’t work in the latest version.

Good luck with it.

**Reply by damiensellier** [reply]
with the latest versions the property values are passed with setProperty() as decimals, still don't know why.
I don't have a midi keyboard with me to test but it would require the usual  **tonumber()** somewhere.


`local ccNumber = tonumber(midi:getData():getByte(1))`
or 

```
local ccNumber = midi:getData():getByte(1)
_c(LEARNINGMOD):getMidiMessage():setProperty("midiMessageCtrlrNumber", tonumber(ccNumber), true)
...
```

Let me know if it's working like this ;)

**Reply by dnaldoog** [reply]
There's also `setPropertyInt(int,bool)` and `setPropertyString(string)` - I wonder if they work on CtrlrX? I mostly always use `setProperty()`

**Comment by dnaldoog** [comment]
**EDIT::** This is an updated version:

@damiensellier  has updated version 5.6.33 to support `setProperty()`.

***Note***: You can use  `_m(LEARNINGMOD):getMidiMessage():seNumber(ccNumber)` but changes are not saved, so when you close the application the MIDI number is not written to the modulator permanently.

<pre>
<code>
midiReceived = function(--[[ CtrlrMidiMessage --]] midi)
    local s = midi:getSize()
    if s == 3 then
        if LEARNINGMOD ~= nil then
            local ccNumber = midi:getData():getByte(1)
            panel:setGlobalVariable(tonumber(LEARNINGMOD:sub(-1)), ccNumber)
            _m(LEARNINGMOD):getMidiMessage():setProperty("midiMessageCtrlrNumber", ccNumber,true)
            _c.statusBar:setText(
                string.format("Set Midi number to : [%.2d] 0x%.2x \187 %s", ccNumber, ccNumber, LEARNINGMOD)
            )
            LEARNINGMOD = nil
        end
    end
end
</code>
</pre>

[MIDI Learn_5_0.zip](https://github.com/user-attachments/files/19968976/MIDI.Learn_5_0.zip)

**Comment by dnaldoog** [comment]
Also I found out that you need:
<pre>
local s = midi:getSize()
    if s == 3 and midi:getType() == 0 then
….
</pre>

Otherwise incoming note messages may be interpreted as cc mapping changes if the user is using a MIDI controller as input.
